core/src/main/java/org/bitcoinj/core/BlockLocator.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2018.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.bitcoinj.core;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class BlockLocator {
+    private ImmutableList<Sha256Hash> hashes;
+    public BlockLocator(){
+        hashes =  ImmutableList.of();
+    }
+    public void add(Sha256Hash hash){
+        hashes = new ImmutableList.Builder<Sha256Hash>().addAll(hashes).add(hash).build();
```

Comment:
and `add` should return a new `BlockLocator`

Reply:
I know that it requires changing the code all around. Wait for @schildbach opinion about it.